src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2StreamWindowManager.cs

```diff
@@ -0,0 +1,257 @@
+            }
+
+            private static TimeSpan StopwatchTicksToTimeSpan(long stopwatchTicks)
+            {
+                long ticks = (long)(StopWatchToTimesSpan * stopwatchTicks);
+                return new TimeSpan(ticks);
+            }
+        }
+
+        private struct RttEstimator
+        {
+            private enum State
+            {
+                Disabled,
+                Init,
+                Waiting,
+                PingSent
+            }
+
+            private const double PingIntervalInSeconds = 1;
+            private static readonly long PingIntervalInTicks =(long)(PingIntervalInSeconds * Stopwatch.Frequency);
+
+            private Http2Connection _connection;
+
+            private State _state;
+            private long _pingSentTimestamp;
+            private long _pingCounter;
+            private int _initialBurst;
+            private volatile bool _terminating;
+            private long _minRtt;
+
+            public TimeSpan MinRtt => new TimeSpan(_minRtt);
+
+            public RttEstimator(Http2Connection connection)
+            {
+                _connection = connection;
+                _state = connection._pool.Settings._disableDynamicHttp2WindowSizing ? State.Disabled : State.Init;
+                _pingCounter = 0;
+                _initialBurst = 4;
+                _pingSentTimestamp = default;
+                _minRtt = 0;
+                _terminating = false;
+            }
+
+            internal void OnInitialSettingsSent()
+            {
+                if (_state == State.Disabled) return;
+                _pingSentTimestamp = Stopwatch.GetTimestamp();
+            }
+
+            internal void OnInitialSettingsAckReceived()
+            {
+                if (_state == State.Disabled) return;
+                RefreshRtt();
+                _state = State.Waiting;
+            }
+
+            internal void OnDataOrHeadersReceived()
+            {
+                if (_state != State.Waiting) return;
+
+                if (_terminating)
+                {
+                    _state = State.Disabled;
+                    return;
+                }
+
+                long now = Stopwatch.GetTimestamp();
+                bool initial = _initialBurst > 0;
+                if (initial || now - _pingSentTimestamp > PingIntervalInTicks)
+                {
+                    if (initial) _initialBurst--;
+
+                    // Send a PING
+                    _pingCounter--;
+                    _connection.LogExceptions(_connection.SendPingAsync(_pingCounter, isAck: false));
+                    _pingSentTimestamp = now;
+                    _state = State.PingSent;
+                }
+            }
+
+            internal void OnPingAckReceived(long payload)
+            {
+                if (_state != State.PingSent)
+                {
+                    ThrowProtocolError();
+                }
+                Debug.Assert(payload < 0);
```

Comment:
Would be helpful to have a message for this assert, since it's not obvious why this must be true.